feat(hook-store): override Hook Gas limit and tokens balances with Tenderly Simulation Data

yvesfracari opened the pull request:
# Summary

Using `gasUsed` data of Tenderly simulation (added on bff on this [PR](https://github.com/cowprotocol/bff/pull/93)) to override `gasLimit` hook data.

# To Test

1. Run the `bff` server locally in this [PR](https://github.com/cowprotocol/bff/pull/93) context
2. Run the app locally connecting it with your local `bff`
3. Use the Build Hook dApp to create one simple hook with a low gas limit. The data that I used to test was a pre-hook with:
- Target: 0xfFf9976782d46CC05630D1f6eBAb18b2324d6B14
- Gas Limit: 10
- Calldata: 0x
4. Select two tokens go until the swap review and check the gas limit. Check if the gas limit is different from the one that you inputted and 10% bigger than the simulated one (click on the Tenderly simulation link to check the simulated)
<img width="327" alt="image" src="https://github.com/user-attachments/assets/5d8cf9e5-71a7-449e-9583-dffbce2a326d">


Note: The PR is in draft until the [balanceDiff PR](https://github.com/cowprotocol/cowswap/pull/5027) is merged. But feel free to test it (only the diffs will have that context as well)

shoom3301 commented on apps/cowswap-frontend/src/modules/hooksStore/containers/HookDappContainer/index.tsx:
```diff
@@ -15,6 +15,7 @@ import { useOrderParams } from '../../hooks/useOrderParams'
 import { HookDapp, HookDappContext as HookDappContextType } from '../../types/hooks'
 import { isHookDappIframe } from '../../utils'
 import { IframeDappContainer } from '../IframeDappContainer'
+import { useHookBalancesDiff } from 'modules/hooksStore/hooks/useBalancesDiff'
```
Nitpick: The import should be relative, since the hooks comes from the same module

shoom3301 commented on apps/cowswap-frontend/src/modules/hooksStore/dapps/BuildHookApp/index.tsx:
```diff
@@ -38,6 +38,10 @@ export function BuildHookApp({ context }: HookDappProps) {
 
+  // JUST FOR TESTING PROPOSES
+  // REMOVE THIS ON MERGE
+  console.log({ balanceDiff: context.balancesDiff })
```
Please remove it

shoom3301 commented on apps/cowswap-frontend/src/modules/hooksStore/hooks/useBalancesDiff.ts:
```diff
@@ -0,0 +1,127 @@
+}
+
+// Returns all the ERC20 Balance Diff of the current hook to be passed to the iframe context
+export function useHookBalancesDiff(isPreHook: boolean, hookToEditDetails?: CowHookDetails): BalancesDiff {
```
Only `uuid` is used here, no need to pass the whole `CowHookDetails`

shoom3301 commented on apps/cowswap-frontend/src/modules/hooksStore/hooks/useBalancesDiff.ts:
```diff
@@ -0,0 +1,127 @@
+    const balanceDiff: Record<string, string> = {}
+
+    if (orderParams?.buyAmount && orderParams.buyTokenAddress && account)
+      balanceDiff[orderParams.buyTokenAddress] = orderParams.buyAmount
```
Should the addresses be lowercased? In order to avoid mixed case issues

shoom3301 commented on apps/cowswap-frontend/src/modules/hooksStore/hooks/useBalancesDiff.ts:
```diff
@@ -0,0 +1,127 @@
+  }, [data, orderParams, preHooks, postHooks])
+}
+
+function mergeBalanceDiffs(
```
Just a shorter version of the code:
```ts
function mergeBalanceDiffs<T = Record<string, Record<string, string>>>(first: T, second: T,): T {
```

shoom3301 commented on apps/cowswap-frontend/src/modules/hooksStore/hooks/useBalancesDiff.ts:
```diff
@@ -0,0 +1,127 @@
+  const result: Record<string, Record<string, string>> = {}
+
+  // Helper function to add BigNumber strings
+  const addBigNumberStrings = (a: string, b: string): string => {
```
This function is pure, so it should be defined outside of the function

shoom3301 commented on apps/cowswap-frontend/src/modules/hooksStore/hooks/useBalancesDiff.ts:
```diff
@@ -0,0 +1,127 @@
+  }
+
+  // Process all addresses from second input
+  for (const address of Object.keys(second)) {
```
Should `address` be lowercased?

shoom3301 commented on apps/cowswap-frontend/src/modules/swap/hooks/useSwapState.tsx:
```diff
@@ -1,6 +1,6 @@
 import { useCallback, useMemo } from 'react'
 
-import { useCurrencyAmountBalance } from '@cowprotocol/balances-and-allowances'
+// import { useCurrencyAmountBalance } from '@cowprotocol/balances-and-allowances'
```
Should be removed

shoom3301 commented on apps/cowswap-frontend/src/modules/swap/hooks/useTokensBalancesCombined.ts:
```diff
@@ -0,0 +1,58 @@
+
+import { usePreHookBalanceDiff } from 'modules/hooksStore/hooks/useBalancesDiff'
+
+export function useTokensBalancesCombined() {
```
The hook should be out of Swap module, since this is generic entity.
Please, create a new module, like `combinedBalances` and move all related hooks there

shoom3301 commented on apps/cowswap-frontend/src/modules/swap/hooks/useCurrencyAmountBalanceCombined.ts:
```diff
@@ -0,0 +1,22 @@
+
+import { useTokensBalancesCombined } from './useTokensBalancesCombined'
+
+export function useCurrencyAmountBalanceCombined(
```
Same as `useTokensBalancesCombined`, the hook should be in a new module

shoom3301 commented on apps/cowswap-frontend/src/modules/tenderly/hooks/useGetTopTokenHolders.ts:
```diff
@@ -0,0 +1,20 @@
+  return useCallback(
+    async (params: GetTopTokenHoldersParams) => {
+      const key = `${params.chainId}-${params.tokenAddress}`
+      if (cachedData[key]?.value) {
```
```ts
return cachedData[key]?.value || fetchTopTokenHolders(params)
```

shoom3301 commented on apps/cowswap-frontend/src/modules/tenderly/hooks/useTenderlyBundleSimulation.ts:
```diff
@@ -0,0 +1,106 @@
+
+  const { data, isValidating: isBundleSimulationLoading } = useSWR(
+    ['tenderly-bundle-simulation', postHooks, preHooks],
+    getNewSimulationData,
```
`preHooks, postHooks` should be passed to the function as arguments.
See:
 - https://github.com/cowprotocol/cowswap/pull/4585
 - https://swr.vercel.app/docs/arguments

shoom3301 commented on apps/cowswap-frontend/src/modules/tokensList/containers/SelectTokenWidget/index.tsx:
```diff
@@ -1,6 +1,6 @@
 import { useCallback, useState } from 'react'
 
-import { useTokensBalances } from '@cowprotocol/balances-and-allowances'
+// import { useTokensBalances } from '@cowprotocol/balances-and-allowances'
```
Remove

shoom3301 commented on apps/cowswap-frontend/src/modules/trade/hooks/useBuildTradeDerivedState.ts:
```diff
@@ -1,13 +1,13 @@
 
 import { Nullish } from 'types'
 
+import { useCurrencyAmountBalanceCombined } from 'modules/swap/hooks/useCurrencyAmountBalanceCombined'
```
`trade` module is global relatively to `swap` and should not depend on it

yvesfracari replied:
Now this is imported from the new `combinedBalances` module. Is this ok or should we import in a different way?

shoom3301 reviewed (changes requested):
---

elena-zh commented:
Hey @yvesfracari , great!
For now, it is the only one open PR related to the Tenderly simulation feature, so I will report issues related to it here: 
1. Simulation report may show a wrong state for a hook. To reproduce: 
- Specify a trade on the Hooks tab
- add a hook (f.e. claim GNO from validators) --> successful simulation
- add 'build your own hook' with some invalid data --> simulation will turn into red (unsuccessful) state for both hooks. 
![failed sim](https://github.com/user-attachments/assets/cdd85617-ac4e-4702-a1a1-9bc0f049061e)
![success](https://github.com/user-attachments/assets/a992e67f-bd82-4738-b8e9-220c23856148)

It would be nice to show correct simulation result for every hook

2. Same happens when a pre-hook is successful one, and a post-hook is not
![post](https://github.com/user-attachments/assets/8a4dcba8-a5bb-4aac-bff8-6151b4f3bf3f)

3. In some cases failed simulation report returns error without a link to the simulation report (see screenshots and cases above). Could you please explain why? 
4. Simulation shows a successful state when I've claimed my funds after the swap (and have nothing to claim). WDYT about re-running simulation again in already added hook when a new trade is specified? 
![claimed](https://github.com/user-attachments/assets/92978a54-0688-4696-87cb-4e6bd80e8bd4)
![success but withdrawn](https://github.com/user-attachments/assets/d670079a-dee1-46f9-a33f-8c447cc9ecc7)
5. When there are more than 2 hooks added, I don't see simulation reports. Could you please check why? 
![several hooks - no simulation](https://github.com/user-attachments/assets/fc62533b-2e32-4b2d-a0c4-c702c061d2ae)


Thanks!

yvesfracari commented:
@shoom3301 

Thanks for all the suggestions. I changed the `BalancesDiff` to be lowercase just after the fetch, so all should already be lowercase everywhere.  Also, implemented the new module as you requested. Could you take another look?

@elena-zh

Thanks for the testing. 

Related to issues 1-3, I think that it is all related to Tenderly call data simulation behavior. To give you more context, before Tenderly runs your simulation it checks if you provided valid call data, this would occur in case of obvious invalid call data (like non-hex decimal string) or invalid data when you compare the provided call data and the contracts abi. If your call data isn't valid, Tenderly doesn't simulate your transaction and just informs you that the transaction is using invalid data. To confirm if you are using valid call data you can try to run the simulation on the Simulator of the Tenderly dashboard (https://dashboard.tenderly.co/ and https://docs.tenderly.co/simulator-ui) and check if it marks your call data as invalid. 

Related to issue 4, we are redoing the simulation on every token or hook change. We don't refresh it on amounts change because it would trigger a large number of simulations due to the quote updates. Every time there is a blue spinner on the hooks it means a new simulation is being made. About this specific issue that you reported, I checked that claiming a vesting contract where you don't have funds anymore does not fail, it `claim` 0 tokens for you, so it is still a successful hook execution. Inside the Claim Hook dApp we're preventing this, however after it is added the simulation will not fail.

Related to issue 5, I couldn't reproduce it, maybe it was already solved on previous code refactoring. Could test again and check if the issue is still happening?

<img width="579" alt="image" src="https://github.com/user-attachments/assets/af8358cf-df05-4dae-9381-534d3131c48e">

yvesfracari commented:
Also @elena-zh fixed here the comment that you did on the other PR (https://github.com/cowprotocol/cowswap/pull/5027#pullrequestreview-2410123926)

elena-zh commented:
Hey @yvesfracari , 

> Related to issues 1-3, I think that it is all related to Tenderly call data simulation behavior. To give you more context, before Tenderly runs your simulation it checks if you provided valid call data, this would occur in case of obvious invalid call data (like non-hex decimal string) or invalid data when you compare the provided call data and the contracts abi. If your call data isn't valid, Tenderly doesn't simulate your transaction and just informs you that the transaction is using invalid data. To confirm if you are using valid call data you can try to run the simulation on the Simulator of the Tenderly dashboard (https://dashboard.tenderly.co/ and https://docs.tenderly.co/simulator-ui) and check if it marks your call data as invalid.

Thank you for explanation. 
But my point here is a bit different one. 
I just wanted to say about the UI issue related to case when: **already added valid hook is labeled as an `invalid (with failed simulation report) one` when a newly added hook's simulation fails.**

This should not behave like this. When a simulation fails for this or that hook, the failed simulation report/state should be displayed for the **appropriate hook only**, not for all hooks that are present on the form. 

> 

> Related to issue 5, I couldn't reproduce it, maybe it was already solved on previous code refactoring. Could test again and check if the issue is still happening?

Yes, it is. 
To reproduce: 
1. On Gnosis chain, add at least 2 prehooks
2. As a post-hook, add 2-3 'Build your own' hooks. Maybe one of them should have invalid data --> see the image
![image](https://github.com/user-attachments/assets/c12ec07a-13d0-4d6c-b285-e3d8b9c7b675)


> Also @elena-zh fixed here the comment that you did on the other PR ([#5027 (review)](https://github.com/cowprotocol/cowswap/pull/5027#pullrequestreview-2410123926))

This case has been fixed, thank you.

yvesfracari commented:
@elena-zh 

Thanks for the report.

Related with issues 1-3:
The tenderly api is kind trick so I am not sure if we can achieve this UX. We’re not simulating individual hooks, but simulating all hooks in a bundle. If one of that hooks fails but with valid calldata (to test this you can create multiple "permit token hooks" with the same token and spender address) the tenderly API will inform to us what hook failed and we can label only the hook that failed as invalid (and also present the tenderly simulation to show more information about the error). However, if some hook has invalid calldata, the Tenderly API just doesn’t simulate anything and return to us an invalid input error. This means that for this kind of error we can’t know which hook failed or if the previous successful hook is still valid or not (since it becomes invalid because of swap or other hooks changes).

This should only happen with this specific kind of data related with invalid input which usually is due to issues while encoding the hook data, invalid address or something like this. 

I talked with @shoom3301 @alfetopito @anxolin and there are plannings to refactor the "build your own hook" dapp to provide have the option of load the contract abi to encode the data proggramatically. Also, I will ask tenderly about if they can change their API to simulate the other hooks even if one of them has some invalid data.

In that sense, I think that this hook is easy to happen on the “build your own hook dapp” since you manually encode the data, however for the other ones it should’t happen since the hooks are encoded programmatic. wdyt?

Related to issue 5:

Thanks for catching this! I think that was a bug related on adding post hook orders without setting the order. This is because i had added a rule that we couldn't simulate post hooks if the order wasn't filled. Now, I changed it to simulate the hook without the order output.

elena-zh reviewed (approved):
All is good rn besides the issue (edge-case) we decided to address later. 
Thanks!

anxolin commented on apps/cowswap-frontend/src/modules/combinedBalances/hooks/useTokensBalancesCombined.ts:
```diff
@@ -0,0 +1,60 @@
+  }, [account, preHooksBalancesDiff, tokenBalances, isHooksTradeType])
+}
+
+function applyBalanceDiffs(currentState: BalancesState, diffs: Record<string, string>): BalancesState {
```
nit: `currentBalances` not current state. Also `balanceDiff` instead of `diffs`

anxolin commented on apps/cowswap-frontend/src/modules/combinedBalances/hooks/useTokensBalancesCombined.ts:
```diff
@@ -0,0 +1,60 @@
+  // Normalize diffs
+  for (const address of Object.keys(diffs)) {
+    normalizedDiffs[address] = diffs[address]
+  }
```
nitpick: it fees a bit more complex than needed. Would something in this line work?

```js
const balanceEntries = concurrentState.values.entries().concat(diff.entries)
return balanceEntries.reduce(([address, balance]) => {
   ...
)
````

anxolin commented on apps/cowswap-frontend/src/modules/hooksStore/hooks/useBalancesDiff.ts:
```diff
@@ -0,0 +1,124 @@
+  const { preHooks, postHooks } = useHooks()
+  const preHookBalanceDiff = usePreHookBalanceDiff()
+
+  const orderMockBalanceDiff = useMemo(() => {
```
why mock?

yvesfracari replied:
Sorry for the lack of documentation. This is used mainly when the swap isn't simulated (like if the user has only preHooks). I changed a bit the variable name and comments.

anxolin replied:
I see, thanks

anxolin commented on apps/cowswap-frontend/src/modules/hooksStore/hooks/useBalancesDiff.ts:
```diff
@@ -0,0 +1,124 @@
+  const bigA = BigNumber.from(a)
+  const bigB = BigNumber.from(b)
+  return bigA.add(bigB).toString()
+}
```
this looks like a utility function that should be somewhere else?

yvesfracari replied:
I think that this utility was so simple and just used once that we don't need it actually 😅

anxolin commented on apps/cowswap-frontend/src/modules/hooksStore/hooks/useBalancesDiff.ts:
```diff
@@ -0,0 +1,124 @@
+  return bigA.add(bigB).toString()
+}
+
+function mergeBalanceDiffs(first: BalancesDiff, second: BalancesDiff): BalancesDiff {
```
Checking the type of the `BalancesDiff` I see is a structure to hold token balances. Having `Diff` in its name feels a bit constraining cause you can use it for example to hold the current balances, etc.

I would drop the `diff` from the name and would make the merging utilities more generic so they live not in the `hookStore` but in a more central place. Or at least can be moved to a utility instead of being in a hook file

yvesfracari replied:
I think the main difference between this and a token balances is that it can be negative. Another difference between this and the token balances used on the rest of the app is the first key. Another difference is that the outer key of `BalanceDiff` is the `address` which the difference is related to while the `BalanceState` of the rest of the app the outer key is the `chainId` since it is always related to the connected account. For now, I don't see any other place on the app where we would store balances diff from addresses, so I think that there isn't too much gain of moving this to a utility for now.

anxolin replied:
ok, makes sense

anxolin commented on apps/cowswap-frontend/src/modules/tenderly/utils/generateSimulationData.ts:
```diff
@@ -18,16 +18,40 @@ export function generateSimulationDataToError(
 
+function convertBalanceDiffToLowerCaseKeys(data: BalancesDiff): BalancesDiff {
+  return Object.entries(data).reduce((acc, [outerKey, innerObj]) => {
+    const lowerOuterKey = outerKey.toLowerCase()
```
why inner/outer? what are these keys? Is it user account and token address? why not stating that in the name?

yvesfracari replied:
Changed the names of the variables for a better documentation!

anxolin reviewed (commented):
The description of this PR made me think the scope is smaller. I don't think the description and code changes maches. Most of this PR is about calculating the aggregated balance given the current user's balance and the result from the similation.

yvesfracari commented on apps/cowswap-frontend/src/modules/hooksStore/hooks/useBalancesDiff.ts:
```diff
@@ -0,0 +1,124 @@
+            result[address][token] = addBigNumberStrings(result[address][token], second[address][token])
+          } catch (error) {
+            console.error(`Error adding balances for address ${address} and token ${token}:`, error)
+            throw error
```
I think that is similar, however is used for only one value instead of all of them and that is why I am not importing the function from there.

yvesfracari commented:
Thanks for the review @anxolin ! I tried to resolve / reply all your questions. About the PR size, sorry it was supposed to be two separated PRs but then it was requested to be merged and I forget to update the description.

About the gas limit, I already faced some issues of hooks not being executed due to out of gas, but never when the gas was defined by the tenderly. About the 10%, I just keep an old logic that existed on other hooks.

yvesfracari commented:
As discussed on the channel, I removed the buffer of the Tenderly simulation gas.

yvesfracari commented:
> **1** out of **2** committers have signed the CLA.✅ @yvesfracari❌ @pedro Yves**Pedro Yves** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).You can retrigger this bot by commenting **recheck** in this Pull Request

This was due some commits with a wrong author, already fixed it. Please ignore it 😅

elena-zh commented:
Changes LGTM.
Not for the current PR, but for the future updates, it would be nice to have one generic 're-run simulation' button that will retrigger bundle on click instead of having several ones with the same action :) Something link this:
![image](https://github.com/user-attachments/assets/05276311-83ca-4beb-b4d0-456ab19c75eb)

@fairlighteth, WDYT? :)

anxolin commented:
@yvesfracari I think you have some lint issues:

<img width="1157" alt="image" src="https://github.com/user-attachments/assets/234d182b-bcb8-479b-bbf6-ed9c75644ea3">

anxolin commented:
I want to merge this PR so we start to consolidate things. Any enhancement we can try to do iterativelly to not keep this one open for too long
